---
title: "DESeq2_lung"
output:
  html_notebook:
    toc: yes
    theme: united
---
https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html
This is the analysis of the MAIT project using DESeq2.



1. setup

```{r, eval = FALSE}
## ----setup, echo=FALSE, results="hide"----------------------------------------
knitr::opts_chunk$set(tidy = FALSE,
                      cache = FALSE,
                      dev = "png",
                      message = TRUE,
                      error = FALSE,
                      echo = TRUE,
                      warning = TRUE)
```
```{r}
library("DESeq2")
library("vsn")
library("hexbin")
library("pheatmap")
library("RColorBrewer")
library("ggplot2")
library("tidyverse")
```
Making it possible to run both on local machine as well as on cluster
```{r}
locale <- FALSE
recompute <- TRUE

prefix <- ifelse(locale, "~", "")
dir <- file.path(prefix, "data/bioinf/projects/data/2022_MLseq_KP/MAIT_DESeq2")
data_dir <- file.path(dir, "data/")
result_dir <- file.path(dir, "results/")
graphics_dir <- file.path(dir, "graphics/")
```
2. load and read file
Roughly following the script https://git.ufz.de/rnaseq/2018_maitcellsrnaseq/-/blob/main/DE_Analysis/01_DESeq2.Rmd
```{r}
count_files <- list.files(path=data_dir, pattern="MAIT_16h", full.names=TRUE)
names(count_files) <- gsub("-htseq_counts.txt", "", basename(count_files)) # this probably shortens the names

SampleTableFile <- paste0(data_dir, "SampleTable_MAITseq_All.csv")
backgroundInfo <- read.csv(SampleTableFile,  sep=",") #info about gender, age etc



file_sampleTable <- file.path(result_dir, "01_sampleTable")

if(recompute) {
    STcolsOfInterest <- c(1:16)
    STcolsAsFactors <- c(4,8,9,12,16)

    sT <- read.table(SampleTableFile, header=T, sep=",", stringsAsFactors = FALSE)
    for(i in STcolsAsFactors)
    { #select requested columns
      sT[,i] <- as.factor(sT[,i])
    }
    sT <- sT[,STcolsOfInterest]
    sampleTable <- sT
    colnames(sampleTable)[c(2,4,5)] <- c("sampleName", "time", "condition")
    rownames(sampleTable) <- sampleTable$sampleName
    sampleTable$time <- as.factor(paste0(sampleTable$time, "h"))
    # fileName is required by DESeq2 function
    rows <- match(names(count_files), rownames(sampleTable))
    sampleTable$fileName[rows] <- basename(count_files)
    # formatting condition etc.
    sampleTable$condition <- ifelse(sampleTable$condition=='-', "Ctrl_0",
                                    ifelse(sampleTable$condition=='unstim',
                                           "Ctrl_16", sampleTable$condition))
    sampleTable$condition <- gsub("-", "_", sampleTable$condition)
    sampleTable$condition <- as.factor(sampleTable$condition)
    sampleTable$treatment <- ifelse(str_detect(sampleTable$condition,"Ctrl"),
                                    "control", "treated")
    sampleTable$treatment <- as.factor(sampleTable$treatment)
    sampleTable <- sampleTable[,c(2,17,4,5,18,3,6:16)]
    save(sampleTable, file=file_sampleTable)
} else {
    load(file = file_sampleTable)
}
print(paste("Sample table loaded:", SampleTableFile, "-",
            dim(sampleTable)[2], "columns for",
            dim(sampleTable)[1], "samples:"))
sampleTable$condition <- gsub("CD3_28", "CD3-28",
                              gsub("Ctrl_16", "16h_Ctrl",
                                   gsub("Ctrl_0", "0h_Ctrl", sampleTable$condition)))
names(sampleTable)


dds <- DESeqDataSetFromHTSeqCount(sampleTable=sampleTable,
                                  directory= data_dir,
                                  design=design)
#look at object
dds

```
3. Pre-filtering
by excluding low count genes we reduce the memory usage of the later methods
```{r}
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
```
This reduced genes from 20500 to 20100 -> not very impressive

4. set factor levels
It may be important to define in which order the conditions are compared. (default is alphabetical order)
```{r}
dds$condition <- factor(dds$condition, levels = c("T1","T2"))
```
The following steps are to visualize and cluster the data. For this process a transformed version of the data is useful.
5. rlog-transformation
You do this if you don't want to do DE but visualization and clustering instead. Main point is to remove the high variance for low means
rlog ~ regularized logarithm - normalizes on basis of library size and transforms data to log2 scale
Quote from DESeq2 paper: we use the shrinkage approach of DESeq2 to implement a regularized logarithm transformation (rlog), which behaves similarly to a log2 transformation for genes with high counts, while shrinking together the values for different samples for genes with low counts. It therefore avoids a commonly observed property of the standard logarithm transformation, the spreading apart of data for genes with low counts, where random noise is likely to dominate any biologically meaningful signal.
(For a big number of samples there is an alternative: VST ~  variance stabilizing transformation)
```{r}
file_vsd <- file.path(result_dir, "01_vsd.Rdata")
file_rld <- file.path(result_dir, "02_rld.Rdata")
if (recompute) {
  vsd <- varianceStabilizingTransformation(dds, blind = FALSE)
  rld <- rlog(dds, blind = FALSE)
  save(vsd, file = file_vsd)
  save(rld, file = file_rld)
} else {
  load(file = file_vsd)
  load(file = file_rld)
}
head(assay(rld), 3)

```
6. Plot showing dependence of sd on mean
```{r}
#plot of sd and mean
ntd <- normTransform(dds)

jpeg(file = file.path(graphics_dir, "01_ntd_meanSd.jpeg"))
meanSdPlot(assay(ntd))
dev.off()
jpeg(file = file.path(graphics_dir, "02_rld_meanSd.jpeg"))
meanSdPlot(assay(rld))
dev.off()
jpeg(file = file.path(graphics_dir, "03_vsd_meanSd.jpeg"))
meanSdPlot(assay(vsd))
dev.off()
```
rld has the most even mean
7. Heatmap
We need to first estimate the size factors to make it work
```{r}

dds <- estimateSizeFactors(dds)
select <- order(rowMeans(counts(dds, normalized = TRUE)),
                decreasing = TRUE)[1:20]
df <- as.data.frame(colData(dds)[, c("condition")])
#jpeg(file="/home/pilz/data/graphics/heatmap.jpeg")
#saving the pheatmap is complex
heatmap <- pheatmap(assay(rld)[select,], cluster_rows = FALSE, show_rownames = FALSE,
                    cluster_cols = FALSE, annotation_col = df)
save_pheatmap_pdf <- function(x, filename, width = 7, height = 7) {
  stopifnot(!missing(x))
  stopifnot(!missing(filename))
  pdf(filename, width = width, height = height)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}
save_pheatmap_pdf(heatmap, file.path(graphics_dir, "04_rld_heatmap.pdf"))
#dev.off()
```
8. Heatmap of the sample-to-sample distances
sampleDists gives distances
creating plot with RColorBrewer and pheatmap
```{r}
sampleDists <- dist(t(assay(rld)))
#plot

sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(rld$condition)
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")))(255)
heatmap2 <- pheatmap(sampleDistMatrix,
                     clustering_distance_rows = sampleDists,
                     clustering_distance_cols = sampleDists,
                     col = colors)
save_pheatmap_pdf(heatmap2, file.path(graphics_dir, "05_rld_heatmap_sampledist.pdf"))
```
9. PCA
Shows samples in 2D plot, the axes representing their two principal components
```{r}

jpeg(file=file.path(graphics_dir, "06_rld_PCA.jpeg"))
plotPCA(rld) + geom_text(aes(label = name), vjust = 2)
dev.off()
```
Observation: Sample 15 is a clear outlier on the y axis
Exclude sample 15 and redo PCA
```{r}
#maybe an alternative:
#keep <- rowSums(counts(dds)) >= 10
#dds <- dds[keep,]

rld <- rld[, -15]
dds <- dds[, -15]

jpeg(file = file.path(graphics_dir, "07_rld_PCA.jpeg"))
plotPCA(rld) + geom_text(aes(label = name), vjust = 2)
dev.off()
```
Obsvation: Sample 54 is a clear outlier
```{r}
rld <- rld[, -53] #since a sample before it was excluded, 54 is in position 53
dds <- dds[, -53]

jpeg(file = file.path(graphics_dir, "08_rld_PCA.jpeg"))
plotPCA(rld) + geom_text(aes(label = name), vjust = 2)
dev.off()
```
Now, 29 also looks like an outlier.
```{r}
rld <- rld[, -28] #again, 29 is now at position 28 because 15 was excluded
dds <- dds[, -28]

jpeg(file = file.path(graphics_dir, "08_rld_PCA.jpeg"))
plotPCA(rld) + geom_text(aes(label = name), vjust = 2)
dev.off()
```
No obvious outliers anymore. Clustering is completed.

13. Standard differential expression analysis
This method goes through all steps of DESeq2 and gives results.
```{r}
file_res <- file.path(result_dir, "03_res.Rdata")
file_dds <- file.path(result_dir, "04_dds.Rdata")
if (recompute) {
  dds <- DESeq(dds)
  res <- results(dds)
  save(dds, file = file_dds)
  save(res, file = file_res)
} else {
  load(file = file_dds)
  load(file = file_res)
}

res
```
12. Perform LFC transformation
For better visualization and gene ranking we use the method lfcShrink
```{r}
resultsNames(dds)
resLFC <- lfcShrink(dds, coef = "condition_treated_vs_untreated", type = "apeglm")
resLFC
```
Error. resultsNamesDDS is not TRUE
13. Table ordered by p-value
For a better overview we can order the table by p-value and give a summary of how many p-values were <0.1
Finally, we can get a results table.
```{r}
resOrdered <- res[order(res$pvalue),]
resOrdered
summary(res)
#How many below 0.1?
sum(res$padj < 0.1, na.rm = TRUE)
#results table
res05 <- results(dds, alpha = 0.05)
summary(res05)
sum(res05$padj < 0.05, na.rm = TRUE)
```
(optional) Independent hypothesis weighting
14. MA-plot
```{r}
plotMA(res, ylim = c(-2, 2))
```
This looks nothing like what I expected. I looks like it is mirrored by the y-axis and moved -1.6 into x-direction

Same plot for shrunken log2 fold chages (removes noise)
```{r}
plotMA(resLFC, ylim = c(-2, 2))
```
Does not work due to Error in 12
15. Plot counts for individual genes
```{r}
plotCounts(dds, gene = which.min(res$padj), intgroup = "condition")
```
16. Export the list of genes that pass p-value
Why can I not run this code chunk?
```{r}
resSig <- subset(resOrdered, padj < 0.1)
resSig
write.csv(as.data.frame(resSig), file = "results_cervical_p01_no_exlclusion")
```