
---
title: "Human MAITcells - RNAseq DE Analysis"
author: "Jana Schor"
date: "Dec 13 (Apr 11), 2018"
output:
  html_document:
    toc: true
    theme: united
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Parameters

<p align="right"><a href="#top">&uarr; Back to top</a></p>

## Project description

<p align="right"><a href="#top">&uarr; Back to top</a></p>

This project covers the differential expression (DE) analysis of MAITcell transcriptome data. The
MAIT cells have been sorted and isolated from human BuffyCoats - all male. Details on the sequencing
experiment can be found on our wikipage:

https://wiki.ufz.de/yigbt/index.php/2018_MAITseq

## Samples and sample groups

5 replicates for each of the 4 conditions (sorted MAIT T-cells isolated from
human BuffyCoats - all male)

- Condition 1: 0h
- Condition 2: 16h_CD3/28_stim, anti body that should result in an immune
  reaction
- Condition 3: 16h_K12_stim, E.coli bacteria that also should result in an
  immune reaction
- Condition 4: 16h_unstim , untreated control, effects might be explained by
  time only

## Research question

0. Assess the data quality and sample to trait association.
  - Principle component and sample-to-sample analysis of the input data
1. Identify differentially expressed genes that may play a role in the
   activation/stimulation of MAIT cells with ecoli (K12) or CD3/28
2. Annotate the most relevant novel transcritps in this context

Identify differentially expressed genes between treated and untreated cells.
In particular study the following comparisons:

A. effect of time (2B subtracted from following comparisons): Cond1 vs Cond4 B.
CD3-28 treated vs untreated (Cond2 vs Cond4) C. K12 treated vs untreated (Cond3
vs Cond4) D. CD3-28 treated vs. K12 treated (Cond2 vs Cond3) compare 2
intersection of B and C


# Environment settings

<p align="right"><a href="#top">&uarr; Back to top</a></p>

## Global environment

<p align="right"><a href="#top">&uarr; Back to top</a></p>

This project is done in the conda environment `DEA18`.
The respective `DEA18_environment.yml` from the git-repo can be used to (re-)create this environment.

```{r create_conda_env, engine='bash', eval=FALSE}
# CMD to create the environment, NOTE: the individual releases might change over time
conda create -n DEA19 -c conda-forge -c bioconda r-base=3.5.1 r-cairo r-tidyverse r-rcolorbrewer r-knitr r-pheatmap r-markdown r-gridextra r-factominer r-factoextra pandoc r-wgcna bioconductor-deseq2 bioconductor-biomart bioconductor-egsea bioconductor-egseadata bioconductor-biocparallel
# Create exactly the same environment from the environment file
conda env create -f DEA19_environment.yml
# Activate the environment
conda activate DEA19
# Start R or rstudio
R # or rstudio
```

This `.Rmd` file that is the basis of this report has been rendered via:

```{r render, eval=FALSE}
rmarkdown::render("01_DESeq2.Rmd", output_format="html_document")
```

## Local (R) environment

<p align="right"><a href="#top">&uarr; Back to top</a></p>

Here, all directories are set to the correct paths of this project. Depending on
the machine that is used, the number of CPUs are adjusted.

```{r set_directories}
# working directory
local <- 0
if(local) {
    cores2Bused <- 3
} else {
    cores2Bused <- 12
}

# my_wd <- "~/git-hertelj/projects/2018_MAITcellsRNAseq/DE_Analysis/"
data_dir <- "/data/bioinf/projects/data/2018_MAITcellsRNAseq/DE_Analysis/"
#setwd(my_wd)
# for storing .Rdata files
RdataDir <- paste0(data_dir, "RData/")
# for storing all result figures, tables, etc.
#outdir <- paste0(data_dir, "results/")
# for sourcing own helper scripts
#sourcedir <- "~/git-hertelj/code/R-helpers/"
```

All necessary libraries are installed/loaded:

```{r load_DESeq2, message=FALSE, warning=FALSE, results='hide'}
# my helper functions for diverse tasks in this project
source(paste0(sourcedir, "myRhelper.R"))
# my helper functions wrapping DESeq2 functions
source(paste0(sourcedir, "myDESeq2helper.R"))
# the following packages are needed for this analysis:
Bioconductor_package_list <- c("biomaRt", "DESeq2")
R_package_list <- c("RColorBrewer", "pheatmap", "BiocParallel",
                    "knitr", "xtable", "tidyverse")
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("biomaRt", version = "3.8")

# installation/loading process of packages
packages <- c(Bioconductor_package_list, R_package_list)
if(length(packages) > 0) {
    load_libraries(c(Bioconductor_package_list, R_package_list), verbose = TRUE)
}
# register multiple cores for processing
mcparam <- MulticoreParam(cores2Bused)
register(mcparam)
```

The following switch allows for faster report generation, if the respective
calculations have been done already. If you want/need to recompute set the
`recompute` variable to 1.

```{r recompute}
recompute <- 0
```

Define a custom minimal theme without background and grid

```{r defThemeBasic}
theme_basic <- function() {

    theme_minimal() %+replace%
        theme(
            panel.grid = element_blank(),
            complete = TRUE
        )
}
```

# Input data

<p align="right"><a href="#top">&uarr; Back to top</a></p>

Set the directory where the **Read Counts** can be found, and the path to the **Sample Table** file.
Define the gene_symbol for the species of this project.

```{r input_file_dir}
countDataDir <- paste0(data_dir, "input/counts")
count_files <- list.files(path=countDataDir, pattern=".txt", full.names=TRUE)
names(count_files) <- gsub("-htseq_counts.txt", "", basename(count_files))
sampleTableFile <- paste0(data_dir, "input/SampleTable_MAITseq_All.csv")
geneSymbol <- "hgnc_symbol" #?
species <- "human" #?
```

## Retreive and transform Sample Table

<p align="right"><a href="#top">&uarr; Back to top</a></p>

The sample table contains additional information about the individual samples.
Select the columns of interest, define which columns should be treated as
factors and retreive the respective table.

```{r get_ST}
file <- paste0(RdataDir, "01_sampleTable.RData")
if(recompute) {
    STcolsOfInterest <- c(1:16)
    STcolsAsFactors <- c(4,8,9,12,16)
    sampleTable <- get.sampleTable(sT_file = sampleTableFile,
                                   sep = ",",
                                   columns = STcolsOfInterest,
                                   factors = STcolsAsFactors)
    colnames(sampleTable)[c(2,4,5)] <- c("sampleName", "time", "condition")
    rownames(sampleTable) <- sampleTable$sampleName
    sampleTable$time <- as.factor(paste0(sampleTable$time, "h"))
    # fileName is required by DESeq2 function
    rows <- match(names(count_files), rownames(sampleTable))
    sampleTable$fileName[rows] <- basename(count_files)
    # formatting condition etc.
    sampleTable$condition <- ifelse(sampleTable$condition=='-', "Ctrl_0",
                                    ifelse(sampleTable$condition=='unstim',
                                           "Ctrl_16", sampleTable$condition))
    sampleTable$condition <- gsub("-", "_", sampleTable$condition)
    sampleTable$condition <- as.factor(sampleTable$condition)
    sampleTable$treatment <- ifelse(str_detect(sampleTable$condition,"Ctrl"),
                                    "control", "treated")
    sampleTable$treatment <- as.factor(sampleTable$treatment)
    sampleTable <- sampleTable[,c(2,17,4,5,18,3,6:16)]
    save(sampleTable, file=file)
} else {
    load(file)
}
print(paste("Sample table loaded:", sampleTableFile, "-",
            dim(sampleTable)[2], "columns for",
            dim(sampleTable)[1], "samples:"))
sampleTable$condition <- gsub("CD3_28", "CD3-28",
                              gsub("Ctrl_16", "16h_Ctrl",
                                   gsub("Ctrl_0", "0h_Ctrl", sampleTable$condition)))
names(sampleTable)
```

## Create the DESeq2 datasets and rm zero-count genes

<p align="right"><a href="#top">&uarr; Back to top</a></p>

We use our count data and the sampleTable to build a **DESeq data set**.
The design is initially set to the just inserted col ```condition```.

```{r build_dds}
# the design formula for the experiment
design <- formula(~ condition)
file <- paste0(RdataDir, "02_dds0.RData")

if(recompute) {
    dds0 <- DESeqDataSetFromHTSeqCount(sampleTable=sampleTable,
                                       directory=countDataDir,
                                       design=design)
    # remove version number of gene names
    rownames(dds0) <- gsub("\\.\\d+", "", rownames(dds0))
    # save this dataframe
    save(dds0, file=file)
} else {
    load(file)
}
dds0
```

We expect that many genes are not expressed at all.
We remove these zero-count genes now from further analysis:

```{r prefiltering}
file <- paste0(RdataDir, "03_dds.RData")
if(recompute) {
    dds <- dds0[ rowSums(counts(dds0)) > 0, ]
    save(dds, file=file)
} else {
    load(file)
}
dds
# save dds as .csv
file <- paste0(outdir, "00_dds_assay_counts.csv")
write.csv2(assay(dds), file=file)
```

## Simplify sample names

```{r simplifySampleNames}
colnames(dds) <- gsub("_stim", "",
                      gsub("unstim", "Ctrl",
                           gsub("MAIT_", "",
                                gsub("0h_", "0h_unstim_", colnames(dds)))))
sampleTable$origName <- rownames(sampleTable)
rownames(sampleTable) <- gsub("_stim", "",
                               gsub("unstim", "Ctrl",
                                    gsub("MAIT_", "",
                                         gsub("0h_", "0h_unstim_", rownames(sampleTable)))))
```

## Retreive gene annotation

<p align="right"><a href="#top">&uarr; Back to top</a></p>

We can get more information about the ENSEMBL genes using R's ```biomaRt```
package. For the new transcripts (XLOCs) information about genomic
location and biotype (= new_transcript) is added to this table.

### Known genes

<p align="right"><a href="#top">&uarr; Back to top</a></p>

```{r get_biomart_table}
file <- paste0(RdataDir, "04_bmTable.RData")
MyAttributes <- c("ensembl_gene_id", "chromosome_name", "start_position",
                  "end_position", "strand", "description",
                  "external_gene_name", "gene_biotype", geneSymbol,
                  "entrezgene")
MyGenesEnsembl <- rownames(dds)[grep("ENS", rownames(dds))]
if(recompute) {
  # we use the same genecode table that we used in the uap anlaysis
  bmTable <- get_biomart_table(Gene_IDs = MyGenesEnsembl,
                               ID_type = "ensembl_gene_id",
                               Gene_attributes = MyAttributes,
                               biomart = "ENSEMBL_MART_ENSEMBL",
                               dataset = "hsapiens_gene_ensembl",
                               host = "may2017.archive.ensembl.org")
  save(bmTable, file = file)
} else {
  load(file)
}
```

- N ENSEMBL genes in data: ```r length(MyGenesEnsembl)```
- N BiomaRt returned genes: ```r dim(bmTable)[1]```

Usually, BiomaRt returns multiple entries for one ID if there are more (other)
IDs available. On the other hand it returns a table that might not contain an
entry for each of our ENSEMBL gene ids.

For repeated (duplicated) entries, the first one is used.

```{r remove_duplicates_bm}
dupl_ens_ids <- bmTable$ensembl_gene_id[duplicated(bmTable$ensembl_gene_id)]
#bmTable <- bmTable[-match(dupl_ens_ids, bmTable$ensembl_gene_id),]
i <- 1
subTable <- bmTable[bmTable$ensembl_gene_id == dupl_ens_ids[i],]
n <- dim(subTable)[1]
rows2Brm <- as.integer(rownames(subTable)[2:n])
for(i in 2:length(dupl_ens_ids)) {
  subTable <- bmTable[bmTable$ensembl_gene_id == dupl_ens_ids[i],]
  n <- dim(subTable)[1]
  rows2Brm <- c(rows2Brm, as.integer(rownames(subTable)[2:n]))
}
rows2Brm <- unique(rows2Brm)
# now remove these secondary entries from the table
bmTable <- bmTable[-rows2Brm,]
```

`NA`s are introduced for not returned genes.

```{r diff_bmTable, results='hide'}
N_missing <- length(MyGenesEnsembl) - length(bmTable$ensembl_gene_id)
if(N_missing > 0) {
    MISSING_ensemble_gene_ids <- MyGenesEnsembl[! MyGenesEnsembl %in%
                                                bmTable$ensembl_gene_id]
    noBMinfo <- c("NA", -1, -1, 0, "No description available", "NA", "NA", "NA", "NA")
    noBMinfo.df <- data.frame(matrix(rep(noBMinfo,
                                         each=length(MISSING_ensemble_gene_ids)),
                                     ncol = length(noBMinfo)))
    m <- as.integer(tail(rownames(bmTable), n=1)) + 1
    mm <- m + length(MISSING_ensemble_gene_ids) -1
    new_rownames <- as.character(c(m:mm))
    MISSING_bm <- data.frame(MISSING_ensemble_gene_ids,
                             row.names = new_rownames, stringsAsFactors = FALSE)
    MISSING_bm <- cbind(MISSING_bm, noBMinfo.df)
    colnames(MISSING_bm) <- colnames(bmTable)
    bmTable.filled <- (rbind(bmTable, MISSING_bm))
} else {
    bmTable.filled <- bmTable
}
```

Now, there should be annotation for all ENSEMBL genes in the data set:

- N ENSEMBL genes in data: ```r length(MyGenesEnsembl)```
- N BiomaRt returned genes: ```r dim(bmTable.filled)[1]```

### New transcripts

<p align="right"><a href="#top">&uarr; Back to top</a></p>

Just for completeness we add all information about genomic location and
description for all newly assembled transcripts (XLOCs) and create a combined
biomaRt table for later use. A .gtf file containing the genes (!) derived from
all cufflinks newly assembled transcripts (exons) has been constructed as
follows:
- input: all XLOCs from merged assembly (cuffcompare results) - this contained exons only
- gffread was applied to merge exons to transcripts
- mergeBed was applied to the transcripts to combine them to genes

See README file in my bioinf2 home at: **/home/hertelj/git-hertelj/projects/2017_LINA/DE_analysis/cufflinks_t2g**

Note: please ensure, that cufflinks and bedtools are available, bevor running this command

```{r get_XLOC_genes, engine='bash', eval=FALSE}
FILE="/data/bioinf/projects/data/2018_MAITcellsRNAseq/uap/filter_class_codes2/mergeGTF-MlE0euH4/mergeGTF-filtered.gtf"
## All entries with XLOC string somewhere on the line
grep -c 'XLOC' $FILE
#159868
## This directly corresponds to the number of XLOC ids:
awk '{print $10;}' $FILE | sed 's/"//g;' | sed 's/;.*$//;' | grep 'XLOC' | wc -l
#159868
## The number of unique XLOC ids is:
awk '{print $10;}' $FILE | sed 's/"//g;' | sed 's/;.*$//;' | grep 'XLOC' | sort -u | wc -l
#150554
## merge exons to transcripts, use gffread
module load gffread/0.10.1
EXTRANS="/data/bioinf/projects/data/2018_MAITcellsRNAseq/DE_Analysis/results/XLOC.exon.transcripts.gtf"
gffread $FILE -o $EXTRANS -F
module unload gffread
## Extract all transcripts (-> discard exon information)
TRANSCRIPTS="/data/bioinf/projects/data/2018_MAITcellsRNAseq/DE_Analysis/results/XLOC.transcripts.gtf"
awk '{if($3=="transcript") print; }' $EXTRANS  | sed 's/\;/\t/g;' | awk '{for(i=1;i<9;i++) printf "%s\t",$i; for(i=9;i<=NF;i++) {if($i ~ /^geneID=/) {printf "%s\n",$i;} } }' > $TRANSCRIPTS
## Check if all XLOC ids are still present or if any have been discarded because they are not part of a transcript or whatever..
awk '{print $NF;}' $TRANSCRIPTS | sort -u | wc -l
#150554
##=> fine.. still all there!
SORTED=${TRANSCRIPTS//\.gtf/\.sorted\.gtf}
sort -k1,1 -k4,5n $TRANSCRIPTS > $SORTED
## Get all IDs (unique list)
IDs="/data/bioinf/projects/data/2018_MAITcellsRNAseq/DE_Analysis/results/XLOC.IDs.lst"
awk '{print $NF;}' $SORTED | sort -u > $IDs
wc -l $IDs
#150554
GENES="/data/bioinf/projects/data/2018_MAITcellsRNAseq/DE_Analysis/results/XLOC.genes.gtf"
for I in $(cat $IDs); do
   C=`grep $I $SORTED | awk '{print $1;}' | sort -u | awk '{printf "%s,",$1;}'`;
   S=`grep $I $SORTED | awk '{printf "%s\n%s\n",$4,$5;}' | sort -u -n | head -1`;
   E=`grep $I $SORTED | awk '{printf "%s\n%s\n",$4,$5;}' | sort -u -n | tail -1`;
   str=`grep $I $SORTED | awk '{print $7;}' | sort -u | awk '{printf "%s,",$1;}'`;
   echo $C"   Cufflinks   gene   "$S"   "$E"   .   "$str"   .   "$I;
done | sed 's/, / /g;' | awk '{for(i=1;i<NF;i++) printf "%s\t",$i; printf "%s\n",$NF;}' > $GENES
## check if all genes are still present
wc -l $GENES
#150554
#sed 's/\+/1/;' $GENES | sed 's/\-/-1/;' > t; mv t $GENES
```

The chunk above has converted all exons to transcripts and merged all
transcripts to genes. From this file I extract the genomic location Using the
generated .gff file to add XLOC information to annotation table.

Still we have some missing xlocs here.. for now i proceed, since this only has
influence on the annotation later..

```{r read_XLOC_gene_gtf}
bmTfile <- paste0(RdataDir, "05_bmTable.combined.RData")
if(recompute) {
    file <-  paste0(outdir, "/XLOC.genes.gtf")
    xloc.info <- read.csv(file, header = FALSE,
                          sep = "\t", stringsAsFactors = FALSE)
    xloc.info <- xloc.info %>%
        mutate('ensembl_gene_id'=gsub("geneID=", "", V9)) %>%
        mutate('chromosome_name'=as.character(V1)) %>%
        mutate('start_position'=as.character(V4)) %>%
        mutate('end_position'=as.character(V5)) %>%
        mutate('strand'=ifelse(V7=="+","1","-1")) %>%
        mutate('description'="Newly assembled transcript using StringTie") %>%
        mutate('external_gene_name'=ensembl_gene_id) %>%
        mutate('gene_biotype'="new_transcript") %>%
        mutate('hgnc_symbol'=ensembl_gene_id) %>%
        mutate('entrezgene'="NA") %>%
        select(c(10:19))
    # test if types are correct
    #sapply(xloc.info, class) == sapply(bmTable.filled, class)
    bmTable.combined <- bind_rows(bmTable.filled, xloc.info)
    save(bmTable.combined, file = bmTfile)
} else {
    load(bmTfile)
}
```

Now we reduce the biotype to a smaller set and define the colors for the
biotypes:

```{r def_my_biotypes}
#biotypes <- unique(bmTable.combined$gene_biotype)
# create a new column 'my_biotype' where I cluster the biotypes to larger groups
bmTable.combined$my_biotype <- bmTable.combined$gene_biotype
bmTable.combined$my_biotype[is.na(bmTable.combined$my_biotype)] <- 'none'
bmTable.combined$my_biotype[grep("pseudogene",
                                 bmTable.combined$my_biotype)] <- "pseudogene"
bmTable.combined$my_biotype[grep("Mt_[tr]RNA",
                                 bmTable.combined$my_biotype)] <- "other"
bmTable.combined$my_biotype[grep("li?ncRNA",
                                 bmTable.combined$my_biotype)] <- "lncRNA"
bmTable.combined$my_biotype[bmTable.combined$my_biotype %in%
                            c("antisense", "processed_transcript", "TEC",
                              "3prime_overlapping_ncRNA", "sense_intronic",
                              "non_coding", "ribozyme", "misc_RNA")] <- "other_ncRNA"
bmTable.combined$my_biotype[bmTable.combined$my_biotype %in%
                            c("snoRNA","miRNA","sRNA", "vaultRNA","snRNA",
                              "scaRNA","scRNA")]  <- "shortRNA"
bmTable.combined$my_biotype[grep("_gene",
                                 bmTable.combined$my_biotype)] <- "protein_coding"
bmTable.combined$my_biotype[! (bmTable.combined$my_biotype %in%
                               c("none","pseudogene","Mt_RNA","lncRNA",
                                 "other_ncRNA", "shortRNA", "protein_coding",
                                 "new_transcript"))] <- "other"
bmTable.combined$reducedBT <- bmTable.combined$my_biotype
bmTable.combined$reducedBT[! (bmTable.combined$reducedBT %in% c("protein_coding", "lncRNA",
                                                                "new_transcript"))] <- "other"

# print overview of biotypes that are present in our data
file <- paste0(RdataDir, "08_bmTable.combined.extended.RData")
save(bmTable.combined, file = file)
file <- paste0(outdir, "00_bmTable.csv")
write.csv2(bmTable.combined, file=file)
```

The resulting biotype and the distribution of the expressed genes over these biotypes are:

```{r table_mybiotype}
tab <- table(bmTable.combined$my_biotype)
df.biotypes <- data.frame(type=names(tab), count=tab %>% as.vector()) %>% arrange(desc(count))
# relevel according to order of counts
df.biotypes$type <- factor(df.biotypes$type,
                      levels=c("protein_coding", "new_transcript", "lncRNA", "shortRNA",
                               "other_ncRNA", "pseudogene", "other"),
                      labels=c("protein-coding", "deNovo-transcript", "lncRNA", "sRNA",
                               "other RNA", "pseudogene", "other"))

p <- ggplot(df.biotypes, aes(x=type, y=count)) +
    geom_bar(stat="identity") +
    theme_basic()

show(p)
```

# General settings

<p align="right"><a href="#top">&uarr; Back to top</a></p>

## Statistics

<p align="right"><a href="#top">&uarr; Back to top</a></p>

Statistical variables for filtering the results

```{r define.stats}
FDR <- 0.01 # FDR, default of DESeq2 is 0.1
p.adjusted <- 0.01
baseMean <- 10 # mean of normalized counts for all samples
```

## Colors

<p align="right"><a href="#top">&uarr; Back to top</a></p>

Beyond the number of differentially expressed genes, we are interested in the
type of gene and its function (if applicable). We color code
- samples according to treatment and time point, and
- genes according to their biotype

```{r def.colors.samples}
# the colors for the samples
# condition -> oranges
colCondition <- brewer.pal(n=4, name='Purples')
names(colCondition) <- c("0h_Ctrl", "16h_Ctrl", "CD3-28", "K12")
# treatment -> reds
colSamples <- c("control" = "#FC9272", "treated" = "#DE2D26")
# time -> blues
colTime <- c("0h" = "#9ECAE1", "16h" = "#3182BD")
l <- length(unique(bmTable.combined$my_biotype))
# the colors for the genes
#colMyBiotypes <- brewer.pal(l, "Set1")
#names(colMyBiotypes) <- unique(bmTable.combined$my_biotype)
# define specific colors for the biotypes
colMyBiotypes <- c(#"other_ncRNA"   = "#23872f", # green
                   #"shortRNA"      = "#41f256", # pale green
                   "protein_coding"= "#42f4ce", # turquoise
                   "lncRNA"        = "#d81e1e", # red
                   #"pseudogene"    = "#1e81d8", # lightblue
                   "new_transcript" = "orange", #"#872323", # darkred
                   "other"    = "gray80" #"#BDBDBD" # grey
)
ann_colors <- list(condition = colCondition,
                   treatment = colSamples,
                   time = colTime,
                   biotype = colMyBiotypes)
```

## Define contrasts

According to the comparisons that should bes considered in the DE analysis,
contrasts are defined as follows:

```{r define_contrasts}
recompute <- 1
file <- paste0(RdataDir, "08_list_contrasts.RData")
if(recompute) {
    list_contrasts <- list(
        # all 16h timepoints
        c_all16=list(dds=dds[,colData(dds)$time == "16h"],
                     design=formula(~ condition),
                     contrast=NULL),
        # all Ctrls 16h vs 0h
        c_allCTRLs=list(dds=dds[,colData(dds)$treatment == "control"],
                        design=formula(~ time),
                        contrast=c("time", "16h", "0h")),
        # 16h Ctrl vs K12
        c_CTRL_K12=list(dds=dds[,colData(dds)$condition == "K12" |
                                             colData(dds)$condition == "16h_Ctrl"],
                        design = formula(~ treatment),
                        contrast=c("treatment", "treated", "control")),
        # 16h Ctrl vs CD3/28
        c_CTRL_CD328=list(dds=dds[,colData(dds)$condition == "CD3-28" |
                                               colData(dds)$condition == "16h_Ctrl"],
                          design = formula(~ treatment),
                          contrast=c("treatment", "treated", "control")),
        # K12 vs CD3/28
        c_K12_CD328=list(dds=dds[,colData(dds)$condition == "K12" |
                                              colData(dds)$condition == "CD3-28"],
                         design = formula(~ condition),
                         contrast=c("condition", "K12", "CD3_28"))
    )
    save(list_contrasts, file=file)
} else {
    load(file)
}
```

# RUN

<p align="right"><a href="#top">&uarr; Back to top</a></p>

## Data quality assessment

<p align="right"><a href="#top">&uarr; Back to top</a></p>

.. by sample clustering and visualization

Our purpose is the detection of differentially expressed genes, and we are
looking in particular for samples whose experimental treatment suffered from an
anormality that renders the data points obtained from these particular samples
detrimental to our purpose.

### Count data transformation

<p align="right"><a href="#top">&uarr; Back to top</a></p>

In order to test for differential expression, we operate on raw counts and use
discrete distributions. However for other downstream analyses e.g. for
visualization or clustering it might be useful to work with transformed versions
of the count data.

The regularized logarithm or rlog is used. It incorporates a prior on the sample
differences and produces transformed data on the log2 scale which has been
normalized with respect to library size. The point of this transformation is to
remove the dependence of the variance on the mean, particularly the high
variance of the logarithm of count data when the mean is low. rlog uses the
experiment-wide trend of variance over mean, in order to transform the data to
remove the experiment-wide trend. Note that we do not require or desire that all
the genes have exactly the same variance after transformation. After the
transformations the genes with the same mean do not have exactly the same
standard deviations, but the experiment-wide trend has flattened. It is those
genes with row variance above the trend which will allow us to cluster samples
into interesting groups.

```{r rlog_transformation}
file <- paste0(RdataDir, "/07_rld.RData")
if(recompute) {
  rld <- rlog(dds, blind=FALSE)
  save(rld, file=file)
} else {
  load(file)
}
# save rld as .csv
file <- paste0(outdir, "01_rld_assay.csv")
write.csv2(assay(rld), file=file)
```

### Sample-to-sample distance

<p align="right"><a href="#top">&uarr; Back to top</a></p>

```{r heatmap_s2s_distance}
sampleDists <- dist(t(assay(rld)))
sampleDistMatrix <- as.matrix(sampleDists)

# annotation of rows
rows <- match(rownames(sampleDistMatrix), rownames(sampleTable))
annotation <- data.frame(condition = sampleTable$condition[rows],
                         time = sampleTable$time[rows],
                         treatment = sampleTable$treatment[rows])
rownames(annotation) <- rownames(sampleDistMatrix)
# heatmap colors
colors <- colorRampPalette( rev(brewer.pal(9, "Greys")) )(255)
# save heatmap in image file
file <- paste0(outdir, "01_heatmap_s2s-distance.pdf")
pheatmap(sampleDistMatrix,
         color=colors,
         border_color = NA,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         treeheight_row = 20,
         treeheight_col = 20,
         legend = T,
         cellwidth=10,
         cellheight=10,
         annotation_row = annotation,
         annotation_col = annotation,
         annotation_colors = ann_colors,
         show_colnames = T,
         main=paste0("Sample-2-sample distance of (rlog-)transformed count data\n",
                     species),
         filename=file)
# print heatmap into report
pheatmap(sampleDistMatrix,
         color=colors,
         border_color = NA,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         treeheight_row = 40,
         treeheight_col = 40,
         legend = T,
         cellwidth=10,
         cellheight=10,
         annotation_row = annotation,
         annotation_col = annotation,
         annotation_colors = ann_colors,
         show_colnames = T,
         main=paste0("Sample-2-sample distance of (rlog-)transformed count data\n",
                     species))
```

Exclude the 0h Ctrl samples and draw the heatmap again:

```{r heatmap_s2s_distance_16}
rld16 <- rld[,grep("16h_", colnames(rld))]
sampleDists16 <- dist(t(assay(rld16)))
sampleDistMatrix16 <- as.matrix(sampleDists16)
# annotation of rows
rows <- match(rownames(sampleDistMatrix16), rownames(sampleTable))
annotation <- data.frame(condition = sampleTable$condition[rows],
                         treatment = sampleTable$treatment[rows])
rownames(annotation) <- rownames(sampleDistMatrix16)
# heatmap colors
colors <- colorRampPalette( rev(brewer.pal(9, "Greys")) )(255)

# save heatmap in image file
file <- paste0(outdir, "01_heatmap_s2s-distance_16.pdf")
pheatmap(sampleDistMatrix16,
         color=colors,
         border_color = NA,
         clustering_distance_rows=sampleDists16,
         clustering_distance_cols=sampleDists16,
         treeheight_row = 40,
         treeheight_col = 0,
         legend = T,
         cellwidth=10,
         cellheight=10,
         annotation_row = annotation,
         annotation_colors = ann_colors,
         show_colnames = T,
         main=paste0("Sample-2-sample distance of (rlog-)transformed count data\n",
                     species, " - 16h only"),
         filename=file)
# print heatmap into report
dev.off()
pheatmap(sampleDistMatrix16,
         color=colors,
         border_color = NA,
         clustering_distance_rows=sampleDists16,
         clustering_distance_cols=sampleDists16,
         treeheight_row = 40,
         treeheight_col = 0,
         legend = T,
         cellwidth=10,
         cellheight=10,
         annotation_row = annotation,
         annotation_col = annotation,
         annotation_colors = ann_colors,
         show_colnames = T,
         main=paste0("Sample-2-sample dist of (rlog-)transformed ",
                     "count data\n",
                     species, " - 16h only"))
```

### Principal component analysis

<p align="right"><a href="#top">&uarr; Back to top</a></p>

Further investigation on potential drivers for the variance in the data is done
via PCA. All columns of the sample table are used as color coding values.

```{r pca, message=FALSE}
PCA_cols <- colnames(colData(dds))
names <- colnames(dds)
plot.PCAs(sampleTable, PCA_cols, names, paste0(outdir, "02_PCA_"), rld, pdfonly=FALSE)
```

Exclude the 0h Ctrl samples and draw the heatmap again:

```{r pca16, message=FALSE}
names16 <- colnames(dds)[grep("16h_", colnames(dds))]
plot.PCAs(sampleTable[grep("16h_", rownames(sampleTable)),],
          PCA_cols, names16, paste0(outdir, "02_PCA_16_"), rld16, pdfonly=FALSE)
```

Remove memory consuming variables.

```{r rm_rld}
rm(list=c('dds', 'dds0',
          'sampleDists', 'sampleDists16',
          'sampleDistMatrix', 'sampleDistMatrix16'))
```


### Observation/Conclustion

<p align="right"><a href="#top">&uarr; Back to top</a></p>

```{r pca_final}
data <- plotPCA(rld, intgroup=c("time", "treatment", "condition"), returnData=TRUE)
percentVar <- round(100*attr(data, "percentVar"))
g <- ggplot(data, aes(x=PC1, y=PC2, color=condition)) +
    geom_point(aes(shape=treatment), size=5) +
    scale_color_manual(values=colCondition) +
    #geom_text(aes(label=names), vjust=2.5, size=4, colour="black") +
    ggplot2::annotate("text", x=-50, y=-15, label="0h Control", size=6) +
    ggplot2::annotate("text", x=-8, y=36, label="16h Control", size=6) +
    ggplot2::annotate("text", x=18, y=-2, label="CD3-28", size=6) +
    ggplot2::annotate("text", x=40, y=-28, label="K12", size=6) +
    labs(title="Main drivers of variance: time + treatment",
         subtitle="Principal component analysis of RNAseq samples, stratified by condition\n",
         caption = "PCA computed based on rlog transformed counts",
         x = paste0("PC1: ", percentVar[1], "% variance"),
         y = paste0("PC2: ", percentVar[2], "% variance")) +
    xlim(-70, 45) +
    ylim(-30,45) +
    theme_bw() +
    theme(
        panel.grid = element_blank(),
        panel.border = element_rect(colour="grey80"),
        plot.title = element_text(face="bold"),
        axis.title.x = element_text(color="black"),
        axis.title.y = element_text(color="black"),
        axis.text = element_text(color="grey80"),
        axis.ticks = element_line(color="grey80"),
        legend.position = "none"
    )
show(g)

file <- paste0(outdir, "X_time_treatment.pdf")
ggsave(g, file=file, width=24, height=24, units="cm")#scale=2)

```

1. Driver of the main variance is time
2. Second driver is the treatment

&Rarr; Design ~condition is right (overall) choice.

&rarr; Design ~time is feasible for comparing the Ctrls only.

&rarr; Design ~treatment is feasible for comparing Ctrl and treated samples for
each treatment separtely.

## DESeq2

<p align="right"><a href="#top">&uarr; Back to top</a></p>


```{r de_analysis_run}
NDE_genesPerContrast <- mat.or.vec(length(list_contrasts), 3)
rownames(NDE_genesPerContrast) <- names(list_contrasts)
colnames(NDE_genesPerContrast) <- c("up", "down", "sum")
c <- 1

lapply(names(list_contrasts), function(name){
    # test
    #name <- names(list_contrasts)[3]
    print(name)

    x <- list_contrasts[[name]]
    # remove zero counts from dds assay
    x$dds <- x$dds[ rowSums(counts(x$dds)) > 0, ]
    # set respective design formula
    design(x$dds) <- x$design

    x$dds$time <- droplevels(x$dds$time)
    x$dds$treatment <- droplevels(x$dds$treatment)
    x$dds$condition <- droplevels(x$dds$condition)

    file_dds <- paste0(RdataDir, "09a_dds_", name, ".RData")
    file_rld <- paste0(RdataDir, "09b_rld_", name, ".RData")

    if(recompute) {
        dds_A <- DESeq(x$dds, parallel=T)
        save(dds_A, file=file_dds)
        # we need this for drawing the heatmap!
        rld_A <- rlog(dds_A, blind=FALSE)
        save(rld_A, file=file_rld)
    } else {
        load(file_dds)
        load(file_rld)
    }

    if(!is.null(x$contrast)) {
        res_unfiltered <- results(dds_A, contrast=x$contrast, parallel=T)
    } else {
        res_unfiltered <- results(dds_A, parallel=T)
    }
    use <- res_unfiltered$baseMean >= baseMean & !is.na(res_unfiltered$pvalue)
    res_filtered <- res_unfiltered[use,]
    # re-djust pvalue
    res_filtered$padj <- p.adjust(res_filtered$pvalue, method="BH")
    # tag differentially expressed genes
    res_filtered$DE <- ifelse(res_filtered$padj <= FDR, TRUE, FALSE)

    ALL_genes <- annotate(res_filtered, bmTable.combined,
                          c(geneSymbol,"description","gene_biotype",
                            "my_biotype", "chromosome_name",
                            "start_position", "end_position",
                            "strand", "entrezgene", "reducedBT"))

    file <- paste0(RdataDir, "09c_res_", name, ".RData")
    save(res_filtered,
         res_unfiltered,
         baseMean,
         FDR,
         file = file)

    # save the DE results for all genes in this comparison
    file <- paste0(outdir, "09d_res_", name, "_anno_ALL.csv")
    write.table(ALL_genes, file=file)
    file <- paste0(RdataDir, "09d_res_", name, "_anno_ALL_genes.RData")
    save(ALL_genes, file=file)

#    N_A2 <- length(rownames(res_A2_filtered))

    # DE genes
    DEgenes<- ALL_genes[ALL_genes$DE,]
    DEgenes$my_biotype[is.na(DEgenes$my_biotype)] <- "new_transcript"

    # save things for DE genes seperately
    file <- paste0(outdir, "09e_res_", name, "_anno_DE.csv")
    write.table(DEgenes, file=file)
    file <- paste0(RdataDir, "09e_res_", name, "anno_DEgenes.RData")
    save(DEgenes, file=file)
    # summary
    summary(res_filtered, alpha=FDR)
    print(table(DEgenes$my_biotype))
    print(table(DEgenes$reducedBT))

    # fill global table
    NDE_genesPerContrast[c,] <<- c(dim(DEgenes[DEgenes$log2FoldChange > 0,])[1],
                                   dim(DEgenes[DEgenes$log2FoldChange < 0,])[1],
                                   dim(DEgenes)[1])

    # plot heatmaps
    if(dim(DEgenes)[1] >= 2) {
        rld_DEgenes <- rld_A[match(rownames(DEgenes), rownames(rld_A)),
                             colnames(dds_A)]

        # annotation of rows (row and colnames are preserved from the rld
        annocol <- as.character(x$design)[2]
        c <- match(annocol, colnames(colData(rld_DEgenes)))
        col <- match(annocol, colnames(colData(rld_DEgenes)))
        col_anno <- data.frame(colData(rld_DEgenes)[col], stringsAsFactors=FALSE)

        # biotypes annotation
        #DEgenes$Biotype = DEgenes$my_biotype
        #DEgenes$Biotype[! (DEgenes$Biotype %in% c("protein_coding",
        #                                          "lncRNA", "new_transcript"))] <- "other"
        row_anno <- data.frame(biotype=factor(DEgenes$reducedBT,
                                              levels=c("protein_coding", "lncRNA", "new_transcript")))
        rownames(row_anno) <- rownames(DEgenes)

        file <- paste0(outdir, "09f_res_DE_heatmap_", name, ".pdf")
#        pdf(file, width=10, height=10, useDingbats=FALSE, onefile=FALSE)

        tmp.assay <- assay(rld_DEgenes)[,match(colnames(rld_DEgenes) %>%
                                               sort(), colnames(rld_DEgenes))]
        pheatmap(tmp.assay,
                 cluster_rows=TRUE,
                 show_rownames=FALSE,
                 treeheight_row = 0,
                 cluster_cols=FALSE,
                 show_colnames=TRUE,
                 treeheight_col = 0,
                 scale='row',
                 main=paste0(dim(DEgenes)[1], " DEgenes (FDR:", FDR, ", ", species, ")\n- Design: ~",
                             as.character(x$design)[2], ", Contr:", name, " -"),
                 fontsize = 8,
                 border_color = NA,
                 annotation_col=col_anno,
                 annotation_row=row_anno,
                 annotation_colors = ann_colors,
                 filename=file
                 )
        # save as .csv in clustered order
        res <- pheatmap(tmp.assay,
                 cluster_rows=TRUE,
                 cluster_cols=FALSE
                 )
        file <- paste0(outdir, "09f_res_DE_heatmap_", name, ".csv")
        hm.csv <- JS_heatmap2csv(assay=tmp.assay, res.hm=res,
                                 row.anno=row_anno, col.anno=col_anno,
                                 file=file)
        # plot into report
        pheatmap(tmp.assay,
                 cluster_rows=TRUE,
                 show_rownames=FALSE,
                 treeheight_row = 0,
                 cluster_cols=FALSE,
                 show_colnames=TRUE,
                 treeheight_col = 0,
                 scale='row',
                 main=paste0(dim(DEgenes)[1], " DEgenes (FDR:", FDR, ", ", species, ")\n- Design: ~",
                             as.character(x$design)[2], ", Contrast:", name, " -"),
                 fontsize = 8,
                 border_color = NA,
                 annotation_col=col_anno,
                 annotation_row=row_anno,
                 annotation_colors = ann_colors
                 )

        dev.off()
    }
    c <<- c + 1
})
file <- paste0(RdataDir, "10_overview_N_DEgenes.RData")
save(NDE_genesPerContrast, file=file)
```

Overview of the number of differentially expressed genes in each contrast:

```{r overviewDE}
kable(NDE_genesPerContrast)
```

Generated result files of this analysis for each contrast are:

```{r overview_files}
list.files(path=outdir, pattern="*", full.names=TRUE)
```

Generated Rdata files of this analysis for each contrast are:

```{r overview_datafiles}
list.files(path=RdataDir, pattern="*", full.names=TRUE)
```

# R session

```{r sessioninfo}
sessionInfo()
```
